File: python/dp/118_Pascal's_Triangle.py
'''
Time: O(n^2) -- n: num of rows
Space: O(n^2)
'''
# temp = [0] + res[-1] + [0] is the key idea
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res = [[1]]  # base case: row 0

        for i in range(numRows - 1):  # already have 1 row
            temp = [0] + res[-1] + [0]  # last row, e.g., [0, 1, 0]
            row = []
            for j in range(len(res[-1]) + 1):
                row.append(temp[j] + temp[j+1])
                
            res.append(row)


        return res

File: javascript/dp/118_Pascal's_Triangle.js
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function(numRows) {
    const res = [[1]]; 

    for (let i=0; i<numRows-1; i++) {
        const temp = [0].concat(res[res.length-1], [0]);  // [0,,] + res[res.length-1] + [,0]
        let row = [];

        for (let j=0; j<res[res.length-1].length+1; j++) row.push(temp[j] + temp[j+1]);

        res.push(row);
    }
    return res;
};